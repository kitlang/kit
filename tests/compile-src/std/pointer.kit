function main() {
    var shared: Shared[CArray[Char, 14]] = Shared.new();
    var sharedString = shared as Ptr[Void] as Shared[CString];

    (*sharedString) = "hello, world!";
    printf("%s\n", *sharedString);
    printf("rc: %zu\n", shared.rc());
    for i in 0 ... 2 {
        printf("incrementing ref\n");
        shared.ref();
        printf("rc: %zu\n", shared.rc());
    }
    for i in 0 ... 3 {
        var released = shared.release();
        printf("released: %s\n", if released then "yes" else "no");
        printf("active: %s\n", if shared.active() then "yes" else "no");
        printf("rc: %zu\n", shared.rc());
    }

    // var s: Shared[Void] = Shared.new();
    // s.release();
}
