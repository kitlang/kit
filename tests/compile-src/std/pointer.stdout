hello, world!
rc: 1
incrementing ref
rc: 2
incrementing ref
rc: 3
released: no
active: yes
rc: 2
released: no
active: yes
rc: 1
released: yes
active: no
rc: 0
